src/datanode/src/sql/copy_table_from.rs

```diff
@@ -0,0 +1,461 @@
+            })?;
+
+        let (path, filename) = DataSource::find_path_and_filename(&path);
+        let source = Source::new(filename, pattern)?;
+
+        let object_store = match result {
+            Some(result) => match result.as_str().to_ascii_uppercase().as_str() {
+                S3_SCHEMA => {
+                    let mut builder = S3::default();
+
+                    builder.root(&path);
+
+                    if let Some(bucket) = host {
+                        builder.bucket(&bucket);
+                    }
+
+                    if let Some(endpoint) = connection.get(ENDPOINT_URL) {
+                        builder.endpoint(endpoint);
+                    }
+
+                    if let Some(region) = connection.get(REGION) {
+                        builder.region(region);
+                    }
+
+                    if let Some(key_id) = connection.get(ACCESS_KEY_ID) {
+                        builder.access_key_id(key_id);
+                    }
+
+                    if let Some(key) = connection.get(SECRET_ACCESS_KEY) {
+                        builder.secret_access_key(key);
+                    }
+
+                    if let Some(session_token) = connection.get(SESSION_TOKEN) {
+                        builder.security_token(session_token);
+                    }
+
+                    if let Some(enable_str) = connection.get(ENABLE_VIRTUAL_HOST_STYLE) {
+                        let enable = enable_str.as_str().parse::<bool>();
+
+                        match enable {
+                            Ok(true) => {
+                                builder.enable_virtual_host_style();
+                            }
+                            Err(_) => {
+                                return Err(error::Error::InvalidConnection {
+                                    msg: format!(
+                                        "failed to parse the {}",
+                                        ENABLE_VIRTUAL_HOST_STYLE
+                                    ),
+                                })
+                            }
+                            _ => (),
+                        }
+                    }
+
+                    let accessor = builder.build().context(error::BuildBackendSnafu)?;
+
+                    ObjectStore::new(accessor).finish()
+                }
+                &_ => {
+                    return Err(error::Error::InvalidUriSchema {
+                        schema: result.as_str().to_string(),
+                    })
+                }
+            },
+            None => {
+                let accessor = Fs::default()
+                    .root(&path)
+                    .build()
+                    .context(error::BuildBackendSnafu)?;
+
+                ObjectStore::new(accessor).finish()
+            }
+        };
+
+        Ok(DataSource {
+            object_store,
+            source,
+            path,
+        })
+    }
+
+    pub async fn list(&self) -> Result<Vec<Object>> {
```

Comment:
Looks like we could hold the regex in DataSource instead of Source. Then `Source` becomes much simpler as it only contains two variants.
```rust
match &self.source {
    Source::Dir => {
        let streamer = self
            .object_store
            .object("/")
            .list()
            .await
            .context(error::ListObjectsSnafu { path: &self.path })?;
        streamer
            .try_filter(|f| {
                let res = if let Some(regex) = &self.regex {
                    regex.is_match(f.name())
                } else {
                    true
                };
                future::ready(res)
            })
            .try_collect::<Vec<_>>()
            .await
            .context(error::ListObjectsSnafu { path: &self.path })
    }
    Source::Filename(filename) => {
        let obj = self.object_store.object(filename);

        Ok(vec![obj])
    }
}
```